File: cpp/population_by_year.cc
// Given a list of people with birth year and death year from 1900 to 2000,
// find the year in which the most people were alive. The algorithm here
// scans the list of people and a list of years once each. Made to work with
// C++98 just for "fun".

#include <stdexcept>
#include <vector>
#include <iostream>

#define FIRST_YEAR 1900
#define LAST_YEAR 2000
#define NUM_YEARS (LAST_YEAR - FIRST_YEAR + 1)

bool valid_year(int year) {
    return year >= FIRST_YEAR && year <= LAST_YEAR;
}

struct Person {
    int birth_year;
    int death_year;

    Person(int by, int dy) : birth_year(by), death_year(dy) {
        if(!valid_year(birth_year) || !valid_year(death_year)) { throw std::out_of_range("bad year"); }
    }
};

std::pair<int, int> find_max(std::vector<Person> people) {
    int years[NUM_YEARS] = {};

    for(std::vector<Person>::iterator it = people.begin(); it != people.end(); ++it) {
        int birth_year_idx = it->birth_year - FIRST_YEAR;
        int death_year_idx = it->death_year - FIRST_YEAR;
        ++years[birth_year_idx];
        --years[death_year_idx];
    }

    int current = years[0], max = years[0];
    int max_year = FIRST_YEAR;
    for(int i = 1; i < NUM_YEARS; ++i) {
        current += years[i];
        if(current > max) {
            max_year = FIRST_YEAR + i;
            max = current;
        }
    }

    return std::make_pair(max_year, max);
}

int main() {
    std::vector<Person> people;
    people.push_back(Person(1901,1903));
    people.push_back(Person(1902,1998));
    people.push_back(Person(1903,1997));
    people.push_back(Person(1904,1996));
    people.push_back(Person(1905,1995));

    std::pair<int, int> max = find_max(people);
    std::cout << "Max of " << max.second << " in " << max.first << std::endl;

    return 0;
}

File: rust/population_by_year.rs
static FIRST_YEAR: i32 = 1900;
static LAST_YEAR: i32 = 2000;
// static NUM_YEARS: usize = (LAST_YEAR - FIRST_YEAR + 1);

struct Person {
  birth_year: i32,
  death_year: i32
}

fn find_max(people: Vec<Person>) -> (i32, i32) {
  let years = [0i32; 101];

  for person in people.iter() {
    let birth_year_idx = person.birth_year - FIRST_YEAR;
    let death_year_idx = person.death_year - FIRST_YEAR;
    years[birth_year_idx] += 1;
    years[death_year_idx] -= 1;
  }

  let (mut current, mut max) = (0i32, 0i32);
  let max_year = FIRST_YEAR;
  for &year in years.iter() {
    current += year;
    if current > max {
      max_year = FIRST_YEAR;
      max = current;
    }
  }

  (max_year, max)
}

fn main() {
  let people = vec![
    Person::new(1901,1903), Person::new(1902,1998), Person::new(1903,1997),
    Person::new(1904,1996), Person::new(1905,1995)
  ];

  let (year, pop) = find_max(people);
  println!("Max of {} in {}", pop, year);
}
